File: src/main/scala/jk_5/nailed/server/map/gamerule/DelegatingGameRules.scala
/*
 * Nailed, a Minecraft PvP server framework
 * Copyright (C) jk-5 <http://github.com/jk-5/>
 * Copyright (C) Nailed team and contributors <http://github.com/nailed/>
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the MIT License.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License
 * for more details.
 *
 * You should have received a copy of the MIT License along with
 * this program. If not, see <http://opensource.org/licenses/MIT/>.
 */

package jk_5.nailed.server.map.gamerule

import java.util

import com.google.common.collect.MapMaker
import jk_5.nailed.api.mappack.gamerule
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.world.GameRules
import org.apache.logging.log4j.LogManager

import scala.collection.convert.wrapAsScala._

/**
 * No description given
 *
 * @author jk-5
 */
object DelegatingGameRules {

  val map = new MapMaker().weakKeys().makeMap[gamerule.GameRules, GameRules]()
  val logger = LogManager.getLogger

  def get(original: gamerule.GameRules): GameRules = {
    if(!map.containsKey(original)){
      val v = new DelegatingGameRules(original)
      map.put(original, v)
      v
    }else{
      map.get(original)
    }
  }
}

class DelegatingGameRules(original: gamerule.GameRules) extends GameRules {

  override def addGameRule(key: String, value: String) = DelegatingGameRules.logger.warn("Tried to add gamerule " + key + " with value " + value + " to immutable gamerules object")
  override def setOrCreateGameRule(key: String, value: String) = DelegatingGameRules.logger.warn("Tried to add gamerule " + key + " with value " + value + " to immutable gamerules object")
  override def getGameRuleStringValue(key: String) = original(key)
  override def getGameRuleBooleanValue(key: String) = original(key) == "true"

  override def writeGameRulesToNBT(): NBTTagCompound = {
    val tag = new NBTTagCompound
    for(rule <- this.original.list){
      tag.setString(rule, original(rule))
    }
    tag
  }

  override def readGameRulesFromNBT(tag: NBTTagCompound) = DelegatingGameRules.logger.warn("Tried to read immutable gamerule object from NBT")
  override def getRules: Array[String] = this.original.list.toArray
  override def hasRule(key: String) = this.original.ruleExists(key)
}

class DelegatingEditableGameRules(private val or: gamerule.EditableGameRules) extends DelegatingGameRules(or) {

  override def addGameRule(key : String, value : String) = or(key) = value
  override def setOrCreateGameRule(key: String, value: String) = or(key) = value
  override def readGameRulesFromNBT(tag: NBTTagCompound){
    val keys = tag.func_150296_c().asInstanceOf[util.Set[String]]
    for(k <- keys){
      this.setOrCreateGameRule(k, tag.getString(k))
    }
  }
}

File: src/main/java/jk_5/nailed/server/map/gamerule/DelegatingGameRules.java
package jk_5.nailed.server.map.gamerule;

import jk_5.nailed.api.gamerule.DefaultGameRuleKey;
import jk_5.nailed.api.gamerule.GameRule;
import jk_5.nailed.api.gamerule.GameRuleKey;
import jk_5.nailed.api.gamerule.GameRules;
import net.minecraft.nbt.NBTTagCompound;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.util.ArrayList;
import java.util.List;

/**
 * No description given
 *
 * @author jk-5
 */
public class DelegatingGameRules extends net.minecraft.world.GameRules {

    private static final Logger logger = LogManager.getLogger();
    private final GameRules<GameRule<?>> wrapped;
    private boolean allowCreation = true;

    public DelegatingGameRules(GameRules<GameRule<?>> wrapped) {
        super();
        this.allowCreation = false; //Creation will be done in the super constructor. Don't log messages when it happens
        this.wrapped = wrapped;
    }

    @Override
    public void addGameRule(String key, String value, ValueType type) {
        if(!allowCreation) logger.warn("Tried to add gamerule {} (value: {}, type: {}) to immutable gamerules object. Ignoring", key, value, type.name());
    }

    @Override
    public void setOrCreateGameRule(String key, String value) {
        logger.warn("Tried to add gamerule {} (value: {}) to immutable gamerules object. Ignoring", key, value);
    }

    @Override
    public String getGameRuleStringValue(String name) {
        return wrapped.get(DefaultGameRuleKey.getByName(name)).toString();
    }

    @Override
    public boolean getGameRuleBooleanValue(String name) {
        GameRuleKey<?> key = DefaultGameRuleKey.getByName(name);
        if(key == null){
            logger.warn("Could not find gamerule key for rule " + name);
            return false;
        }
        if(key.getType() == GameRuleKey.Type.BOOL){
            //noinspection unchecked
            return wrapped.get((GameRuleKey<Boolean>) key).getValue();
        }else{
            logger.warn("Requested boolean gamerule with name {} but it was of type {}", name, key.getType());
            return false;
        }
    }

    @Override
    public int getInt(String name) {
        GameRuleKey<?> key = DefaultGameRuleKey.getByName(name);
        if(key.getType() == GameRuleKey.Type.INTEGER){
            //noinspection unchecked
            return wrapped.get((GameRuleKey<Integer>) key).getValue();
        }else{
            logger.warn("Requested integer gamerule with name {} but it was of type {}", name, key.getType());
            return 0;
        }
    }

    @Override
    public NBTTagCompound writeGameRulesToNBT() {
        NBTTagCompound nbt = new NBTTagCompound();
        for(GameRule<?> rule : wrapped.list()){
            nbt.setString(rule.getKey().getName(), rule.getValue().toString());
        }
        return nbt;
    }

    @Override
    public void readGameRulesFromNBT(NBTTagCompound nbt) {
        logger.warn("Tried to read immutable gamerules object from nbt. Ignoring");
    }

    @Override
    public String[] getRules() {
        List<String> ret = new ArrayList<String>(this.wrapped.list().size());
        for(GameRule<?> rule : this.wrapped.list()){
            ret.add(rule.getKey().getName());
        }
        return ret.toArray(new String[ret.size()]);
    }

    @Override
    public boolean hasRule(String name) {
        for(GameRule<?> rule : this.wrapped.list()){
            if(rule.getKey().getName().equalsIgnoreCase(name)){
                return true;
            }
        }
        return false;
    }

    @Override
    public boolean areSameType(String key, ValueType type) {
        if(type == ValueType.ANY_VALUE) return true;
        GameRuleKey<?> ruleKey = null;
        for(GameRule<?> rule : this.wrapped.list()){
            if(rule.getKey().getName().equalsIgnoreCase(key)){
                ruleKey = rule.getKey();
            }
        }
        if(ruleKey == null) return false;
        return (ruleKey.getType() == GameRuleKey.Type.BOOL && type == ValueType.BOOLEAN_VALUE) || (ruleKey.getType() == GameRuleKey.Type.INTEGER && type == ValueType.NUMERICAL_VALUE);
    }

    @Override
    public String toString() {
        return "DelegatingGameRules{" +
                "wrapped=" + wrapped +
                '}';
    }
}
